XSConsoleData.py

```diff
@@ -528,17 +552,101 @@ def SaveToNTPConf(self):
         # Double-check authentication
         Auth.Inst().AssertAuthenticated()
 
-        file = None
         try:
-            file = open("/etc/chrony.conf", "w")
-            for other in self.ntp.othercontents([]):
-                file.write(other+"\n")
-            for server in self.ntp.servers([]):
-                file.write("server "+server+" iburst\n")
+            with open("/etc/chrony.conf", "w") as confFile:
+                for other in self.ntp.othercontents([]):
+                    confFile.write(other + "\n")
+                for server in self.ntp.servers([]):
+                    confFile.write("server " + server + " iburst\n")
         finally:
-            if file is not None: file.close()
             self.UpdateFromNTPConf()
 
+        # Force chronyd to update the time
+        if self.data['ntp']['method'] != "Disabled":
+            servers = self.data['ntp']['servers']
+            if self.data['ntp']['method'] == "DHCP":
+                servers = self.GetDHClientInterfaces()
+
+            if servers:
+                self.StopService("chronyd")
+                # Single shot time update like: `ntpdate servers[0]`
+                getoutput("chronyd -q 'server %s iburst'" % servers[0])
+                # Write the system time (set by the single shot NTP) to the HW clock
+                getoutput("hwclock -w")
+                self.StartService("chronyd")
+
+    def AddDHCPNTP(self):
+        # Double-check authentication
+        Auth.Inst().AssertAuthenticated()
+
+        oldPermissions = os.stat("/etc/dhcp/dhclient.d/chrony.sh").st_mode
+        newPermissions = oldPermissions | (stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
+        os.chmod("/etc/dhcp/dhclient.d/chrony.sh", newPermissions)
+
+        interfaces = self.GetDHClientInterfaces()
+        for interface in interfaces:
+            ntpServer = self.GetDHCPNTPServer(interface)
+
+            with open("/var/lib/dhclient/chrony.servers.%s" % interface, "w") as chronyFile:
+                chronyFile.write("%s iburst prefer\n" % ntpServer)
+
+    def ResetDefaultNTPServers(self):
+        # Double-check authentication
+        Auth.Inst().AssertAuthenticated()
+
+        Data.Inst().NTPServersSet(
+            [
+                "0.centos.pool.ntp.org",
+                "1.centos.pool.ntp.org",
+                "2.centos.pool.ntp.org",
+                "3.centos.pool.ntp.org",
+            ]
+        )
```

Comment:
After changing this, does it need a call to chrony to update its server list?

Reply:
To make sure the operations are relatively quick and not duplicating any behaviours; most of these functions in XSConsoleData update the underlying config and the kick to chrony occurs in the "Commit" method of XSFeatureNTP.py
I generally wanted to avoid a situation where we partially updated the system before a failure occurred (if any) and left it in an unusable/inconsistent state